File: samples/fifo/Source.cc
//
// This file is part of an OMNeT++/OMNEST simulation example.
//
// Copyright (C) 1992-2015 Andras Varga
//
// This file is distributed WITHOUT ANY WARRANTY. See the file
// `license' for details on this and other legal matters.
//

#include <omnetpp.h>

using namespace omnetpp;

namespace fifo {

/**
 * Generates messages or jobs; see NED file for more info.
 */
class Source : public cSimpleModule
{
  private:
    cMessage *sendMessageEvent = nullptr;

  public:
    virtual ~Source();

  protected:
    virtual void initialize() override;
    virtual void handleMessage(cMessage *msg) override;
};

Define_Module(Source);

Source::~Source()
{
    cancelAndDelete(sendMessageEvent);
}

void Source::initialize()
{
    sendMessageEvent = new cMessage("sendMessageEvent");
    scheduleAt(simTime(), sendMessageEvent);
}

void Source::handleMessage(cMessage *msg)
{
    ASSERT(msg == sendMessageEvent);

    cMessage *job = new cMessage("job");
    send(job, "out");

    scheduleAt(simTime()+par("interarrivalTime").doubleValue(), sendMessageEvent);
}

}; //namespace

File: samples/pyfifo/Source.py
#
# This file is part of an OMNeT++/OMNEST simulation example.
#
# Copyright (C) 1992-2015 Andras Varga
#
# This file is distributed WITHOUT ANY WARRANTY. See the file
# `license' for details on this and other legal matters.
#

from omnetpp.runtime import *

class Source(omnetpp.cSimpleModule):
    """
    Generates messages or jobs; see NED file for more info.
    """

    def initialize(self, stage : int):
        self.sendMessageEvent = omnetpp.cMessage("sendMessageEvent")
        self.scheduleAt(omnetpp.simTime(), self.sendMessageEvent)

    def handleMessage(self, msg : omnetpp.cMessage):
        assert msg == self.sendMessageEvent

        job = omnetpp.cMessage("job")
        self.send(job, "out")
        p = self.par("interarrivalTime")
        self.scheduleAt(omnetpp.simTime() + p.doubleValue(), self.sendMessageEvent)
